qwandaq/src/main/java/life/genny/qwandaq/utils/CapabilityUtils.java

```diff
@@ -165,39 +186,23 @@ public boolean hasCapability(final String rawCapabilityCode, final CapabilityMod
 		}
 		final String cleanCapabilityCode = cleanCapabilityCode(rawCapabilityCode);
 		BaseEntity user = beUtils.getUserBaseEntity();
-
-		// Look through all the user entity attributes for the capability code. If it is
-		// there check the cache with the roleCode as the userCode
-		// TODO: Will need to revisit this implementation with Jasper
-
-		Optional<EntityAttribute> lnkRole = user.findEntityAttribute(LNK_ROLE_CODE);
+		List<String> codes = beUtils.getBaseEntityCodeArrayFromLinkAttribute(user, LNK_ROLE_CODE);
 
 		// Make a list for the modes that have been found in the user's various roles
 		// TODO: Potentially change this to a system that matches from multiple roles
 		// instead of a single role
 		// List<CapabilityMode> foundModes = new ArrayList<>();
 
-		if (lnkRole.isPresent()) {
-			String rolesValue = lnkRole.get().getValueString();
+		for (String code : codes) {
 			try {
-				// Look through cache using each role
-				JsonArray roleArray = jsonb.fromJson(rolesValue, JsonArray.class);
-				for (int i = 0; i < roleArray.size(); i++) {
-					String roleCode = roleArray.getString(i);
-
-					CapabilityMode[] modes = getCapabilitiesFromCache(roleCode, cleanCapabilityCode);
-					List<CapabilityMode> modeList = Arrays.asList(modes);
-					for (CapabilityMode checkMode : checkModes) {
-						if (!modeList.contains(checkMode))
-							return false;
-					}
+				CapabilityMode[] modes = getCapabilitiesFromCache(code, cleanCapabilityCode);
+				List<CapabilityMode> modeList = Arrays.asList(modes);
```

Comment:
Change from List to HashSet to get better performance.

Reply:
Changed to HashSet